psrqpy/search.py

```diff
@@ -18,586 +18,2043 @@
-        self._include_refs = include_refs
-        self._atnf_version = version
-        self._atnf_version = self.get_version  # if no version is set this will return the current or default value
-        self._adsref = adsref
+        XX = np.full(self.catalogue_len, np.nan)
+        YY = np.full(self.catalogue_len, np.nan)
+        ZZ = np.full(self.catalogue_len, np.nan)
 
-        self._savefile = None  # file to save class to
-        self._loadfile = None  # file class loaded from
+        XX[idx] = sc.galactic.cartesian.x.value
+        YY[idx] = sc.galactic.cartesian.y.value
+        ZZ[idx] = sc.galactic.cartesian.z.value
+
+        # set galactic cartesian position
+        self.update(XX, name='XX')
+        self.update(YY, name='YY')
+        self.update(ZZ, name='ZZ')
+
+        # set DMSINB
+        if 'DM' in self.columns:
+            DM = self.catalogue['DM']
+            GB = self.catalogue['GB']
+
+            DMSINB = np.full(self.catalogue_len, np.nan)
+
+            idx = np.isfinite(GB) & np.isfinite(DM)
+            DMSINB[idx] = DM[idx]*np.sin(np.deg2rad(GB[idx]))
+
+            self.update(DMSINB, name='DMSINB')
+
+        # galactic proper motion
+        if not np.all([p in self.columns for p in ['PMB', 'PML']]):
+            reqpar = ['PMRA', 'PMDEC']
+            if np.all([p in self.columns for p in reqpar]):
+                PMRA = self.catalogue['PMRA'].values.copy()
+                PMDEC = self.catalogue['PMDEC'].values.copy()
+
+                PMB = np.full(self.catalogue_len, np.nan)
+                PML = np.full(self.catalogue_len, np.nan)
+
+                idx = (np.isfinite(PMRA) & np.isfinite(PMDEC) &
+                       np.isfinite(RAJD) & np.isfinite(DECJD) &
+                       np.isfinite(DIST))
+
+                sc = ICRS(RAJD[idx]*aunits.deg, DECJD[idx]*aunits.deg,
+                          distance=DIST[idx]*aunits.kpc,
+                          pm_ra_cosdec=PMRA[idx]*aunits.mas/aunits.yr,
+                          pm_dec=PMDEC[idx]*aunits.mas/aunits.yr
+                          ).transform_to(Galactic())
+
+                PMB[idx] = sc.pm_b.value
+                PML[idx] = sc.pm_l_cosb.value
+
+                self.update(PMB, name='PMB')
+                self.update(PML, name='PML')
+
+    def derived_binary(self):
+        """
+        Calculate derived binary system parameters.
+        """
+
+        MASS_PSR = 1.35  # canonical pulsar mass (solar masses)
 
-        if loadfromfile:
-            self.load(loadfromfile)
+        # derive mass function
+        reqpars = ['A1', 'PB']
+        if np.all([p in self.columns for p in reqpars]):
+            A1 = self.catalogue['A1'].values.copy()*c.value
+            PB = self.catalogue['PB'].values.copy()*86400.
+
+            idx = np.isfinite(A1) & np.isfinite(PB)
+
+            MASSFN = np.full(self.catalogue_len, np.nan)
+            MASSFN[idx] = (4.*np.pi**2/GM_sun.value)*A1[idx]**3/(PB[idx]**2)
```

Comment:
I think you might be missing a few factors of c here. The equation should be `f = 4*np.pi**2 * A1[idx]**3 * c**3 / (GM_sun * PB[idx]**2)`

Reply:
> I think you might be missing a few factors of c here. The equation should be `f = (4*np.pi**2 * A1[idx]**3 * c**3) / (GM_sun.value * PB[idx]**2)`

The `A1` values have already been multiplied by c a few lines above.

Reply:
> Also, it might be good to add the error in the mass function. It's not included in the psrcat, but it's useful.

Yes, I would be useful to have. Do you agree that the error should be:

```python
A1ERR = self.catalogue['A1_ERR']*c.value
PBERR = self.catalogue['PB_ERR']*86400.

idx = np.isfinite(MASSFN) & np.isfinite(A1ERR) & np.isfinite(PBERR)

MASSFN_ERR = np.full(self.catalogue_len, np.nan)
MASSFN_ERR[idx] = MASSFN[idx] * np.sqrt( (3.*A1ERR[idx]/A1[idx])**2 + (2.*PBERR[idx]/PB[idx])**2 )
```
I just used standard propagation of errors (assuming `A1` and `PB` are uncorrelated), but it would be good if you could double check.

Reply:
> > I think you might be missing a few factors of c here. The equation should be `f = (4*np.pi**2 * A1[idx]**3 * c**3) / (GM_sun.value * PB[idx]**2)`
> 
> The `A1` values have already been multiplied by c a few lines above.

OK, I didn't see that. It might be good to add a comment about that change of units, though

Reply:
> > Also, it might be good to add the error in the mass function. It's not included in the psrcat, but it's useful.
> 
> Yes, I would be useful to have. Do you agree that the error should be:
> 
> ```python
> A1ERR = self.catalogue['A1_ERR']*c.value
> PBERR = self.catalogue['PB_ERR']*86400.
> 
> idx = np.isfinite(MASSFN) & np.isfinite(A1ERR) & np.isfinite(PBERR)
> 
> MASSFN_ERR = np.full(self.catalogue_len, np.nan)
> MASSFN_ERR[idx] = MASSFN[idx] * np.sqrt( (3.*A1ERR[idx]/A1[idx])**2 + (2.*PBERR[idx]/PB[idx])**2 )
> ```
> 
> I just used standard propagation of errors (assuming `A1` and `PB` are uncorrelated), but it would be good if you could double check.

I think that looks right!

Reply:
Cool, I'll add this in now.